File: Leetcode/Problems C++/1329. Sort the Matrix Diagonally.cpp
//i-j is unique so in one iteration push in map of priority que and in second assign values
//O(MNLog(Diag = min(N,M)))
class Solution {
public:
    vector<vector<int>> diagonalSort(vector<vector<int>>& mat) {
        unordered_map<int,priority_queue<int, vector<int>, greater<int>>>mp;// min priority queue
        for(int i=0;i<mat.size();i++)
        {
           for(int j=0;j<mat[i].size();j++)
           {
               mp[i-j].push(mat[i][j]);
           }
        }        
        
        for(int i=0;i<mat.size();i++)
        {
           for(int j=0;j<mat[i].size();j++)
           {
               mat[i][j] = mp[i-j].top();
               mp[i-j].pop();
           }
        }
        return mat;
    }
};

File: Leetcode/Problems Java/1329. Sort the Matrix Diagonally.java
class Solution {
    public int[][] diagonalSort(int[][] mat) {
        int n = mat.length;
        int m = mat[0].length;
        Map<Integer, PriorityQueue<Integer>> hashmap = new HashMap<>();
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                hashmap.putIfAbsent(i - j, new PriorityQueue<>());
                hashmap.get(i - j).add(mat[i][j]);
            }
        }
        for (int i = 0; i < n; i++)
            for (int j = 0; j < m; j++)
                mat[i][j] = hashmap.get(i - j).poll();


        return mat;
    }
}
